lib/ansible/modules/cloud/vmware/vmware_guest_file_operation.py

```diff
@@ -0,0 +1,379 @@
+    return True
+
+
+def main():
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(dict(datacenter=dict(type='str'),
+                              cluster=dict(type='str'),
+                              folder=dict(type='str'),
+                              vm_id=dict(type='str', required=True),
+                              vm_id_type=dict(default='vm_name', type='str',
+                                              choices=['inventory_path', 'uuid', 'dns_name', 'vm_name']),
+                              vm_username=dict(type='str', required=True),
+                              vm_password=dict(type='str', no_log=True, required=True),
+                              directory=dict(type='dict', default={}),
+                              copy=dict(type='dict', default={}),
+                              fetch=dict(type='dict', default={})
+                              ))
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=False,
+                           required_if=[['vm_id_type', 'inventory_path', ['folder']]],
+                           )
+
+    if not HAS_PYVMOMI:
+        module.fail_json(changed=False, msg='pyvmomi is required for this module')
+
+    if module.params['vm_id_type'] == 'vm_name' and not module.params['folder']:
+        raise Exception('Folder is required parameter when vm_id_type is inventory_path')
+
+    datacenter_name = module.params['datacenter']
+    cluster_name = module.params['cluster']
+    folder = module.params['folder']
+    content = connect_to_api(module)
+
+    datacenter = None
+    if datacenter_name:
+        datacenter = find_datacenter_by_name(content, datacenter_name)
+        if not datacenter:
+            module.fail_json(changed=False, msg="Unable to find %(datacenter)s datacenter" % module.params)
+
+    cluster = None
+    if cluster_name:
+        cluster = find_cluster_by_name(content, cluster_name, datacenter)
+        if not cluster:
+            module.fail_json(changed=False, msg="Unable to find %(cluster)s cluster" % module.params)
+
+    if module.params['vm_id_type'] == 'inventory_path':
+        vm = find_vm_by_id(content, vm_id=module.params['vm_id'], vm_id_type="inventory_path", folder=folder)
+    else:
+        vm = find_vm_by_id(content, vm_id=module.params['vm_id'], vm_id_type=module.params['vm_id_type'],
+                           datacenter=datacenter, cluster=cluster)
+
+    if not vm:
+        module.fail_json(msg='Unable to find virtual machine.')
+
+    try:
+        if len(module.params['directory']) > 0:
+            msg = directory(module, content, vm)
+        if len(module.params['copy']) > 0:
+            msg = copy(module, content, vm)
+        if len(module.params['fetch']) > 0:
+            msg = fetch(module, content, vm)
+        module.exit_json(changed=True, uuid=vm.summary.config.uuid, msg=msg)
```

Comment:
This assumes that we always make changes, even if the directory already exists (or is already missing), or if the file already exists and is not overwritten.